sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java

```diff
@@ -0,0 +1,364 @@
+
+  private void markObjectAtPathAsDead(
+      JsonNode ref, String parentPath, DeadLeafCollector collector) {
+    if (!ref.isObject()) { // it's a scalar
+      collector.addLeaf(parentPath, ImmutableDeadLeaf.builder().name("").build());
+    } else {
+      Iterator<String> fieldNames = ref.fieldNames();
+      while (fieldNames.hasNext()) {
+        String fieldName = fieldNames.next();
+        collector.addLeaf(parentPath, ImmutableDeadLeaf.builder().name(fieldName).build());
+      }
+    }
+  }
+
+  private void markArrayAtPathAsDead(JsonNode ref, String parentPath, DeadLeafCollector collector) {
+    if (!ref.isArray()) { // it's a scalar
+      collector.addLeaf(parentPath, ImmutableDeadLeaf.builder().name("").build());
+    } else {
+      collector.addArray(parentPath);
+    }
+  }
+
+  private void writeLeafIfNewer(
+      JsonNode ref,
+      QueryOuterClass.Row row,
+      String leaf,
+      String parentPath,
+      Map<String, Long> pathWriteTimes,
+      Long rowWriteTime,
+      Map<String, Integer> columnNameToIndex,
+      boolean numericBooleans) {
+    JsonNode n = NullNode.getInstance();
+
+    int txtValueIdx = columnNameToIndex.get("text_value");
+    int boolValueIdx = columnNameToIndex.get("bool_value");
+    int dblValueIdx = columnNameToIndex.get("dbl_value");
+
+    if (!row.getValues(txtValueIdx).hasNull()) {
+      String value = row.getValues(txtValueIdx).getString();
+      if (value.equals(Constants.EMPTY_OBJECT_MARKER)) {
+        n = mapper.createObjectNode();
+      } else if (value.equals(Constants.EMPTY_ARRAY_MARKER)) {
+        n = mapper.createArrayNode();
+      } else {
+        n = new TextNode(value);
+      }
+    } else if (!row.getValues(boolValueIdx).hasNull()) {
+      QueryOuterClass.Value booleanValue = row.getValues(boolValueIdx);
+      Boolean booleanFromRow;
+      if (numericBooleans) {
+        booleanFromRow = booleanValue.getInt() != 0;
+      } else {
+        booleanFromRow = booleanValue.getBoolean();
+      }
+      n = BooleanNode.valueOf(booleanFromRow);
+    } else if (!row.getValues(dblValueIdx).hasNull()) {
+      // If not a fraction represent as a long to the user
+      // This lets us handle queries of doubles and longs without
+      // splitting them into separate columns
+      double dv = row.getValues(dblValueIdx).getDouble();
+      long lv = (long) dv;
+      if ((double) lv == dv) n = new LongNode(lv);
+      else n = new DoubleNode(dv);
+    }
+    if (ref == null)
+      throw new RuntimeException("Missing path @" + leaf + " v=" + n + " row=" + row);
+
+    boolean shouldWrite =
+        !pathWriteTimes.containsKey(parentPath + "." + leaf)
+            || pathWriteTimes.get(parentPath + "." + leaf) <= rowWriteTime;
+    if (shouldWrite) {
+      if (ref.isObject()) {
+        ((ObjectNode) ref).set(leaf, n);
+      } else if (ref.isArray()) {
+        if (!leaf.startsWith("["))
+          throw new RuntimeException("Trying to write object to array " + leaf);
```

Comment:
should we change this and other exceptions with the `ErrorCodeRuntimeException`?

Reply:
This really should be an `IllegalStateException`, as it should never occur normally